app/controllers/incomeSources/manage/ManageIncomeSourceController.scala

```diff
@@ -16,31 +16,85 @@
-                                             val authorisedFunctions: FrontendAuthorisedFunctions,
-                                             val checkSessionTimeout: SessionTimeoutPredicate)
-                                            (implicit val appConfig: FrontendAppConfig,
-                                             mcc: MessagesControllerComponents,
-                                             val ec: ExecutionContext,
-                                             val itvcErrorHandlerAgent: AgentItvcErrorHandler)
-  extends ClientConfirmedController with FeatureSwitching with I18nSupport {
+@Singleton
+class ManageIncomeSourceController @Inject()(val manageIncomeSources: ManageIncomeSources,
+                                             val checkSessionTimeout: SessionTimeoutPredicate,
+                                             val authenticate: AuthenticationPredicate,
+                                             val authorisedFunctions: AuthorisedFunctions,
+                                             val retrieveNino: NinoPredicate,
+                                             val retrieveIncomeSources: IncomeSourceDetailsPredicate,
+                                             val itvcErrorHandler: ItvcErrorHandler,
+                                             implicit val itvcErrorHandlerAgent: AgentItvcErrorHandler,
+                                             val incomeSourceDetailsService: IncomeSourceDetailsService,
+                                             val retrieveBtaNavBar: NavBarPredicate)
+                                            (implicit val ec: ExecutionContext,
+                                             implicit override val mcc: MessagesControllerComponents,
+                                             val appConfig: FrontendAppConfig) extends ClientConfirmedController
+  with FeatureSwitching {
 
-  def show(): Action[AnyContent] = Action {
-    Ok("")
+  def show(): Action[AnyContent] = (checkSessionTimeout andThen authenticate andThen retrieveNino
+    andThen retrieveIncomeSources andThen retrieveBtaNavBar).async {
+    implicit user =>
+      handleRequest(
+        sources = user.incomeSources,
+        isAgent = false,
+        backUrl = controllers.routes.HomeController.show().url
+      )
   }
 
-  def showAgent(): Action[AnyContent] = Action {
-    Ok("")
+  def showAgent(): Action[AnyContent] = Authenticated.async {
+    implicit request =>
+      implicit user =>
+        getMtdItUserWithIncomeSources(incomeSourceDetailsService) flatMap {
+          implicit mtdItUser =>
+            handleRequest(
+              sources = mtdItUser.incomeSources,
+              isAgent = true,
+              backUrl = controllers.routes.HomeController.showAgent.url
+            )
+        }
   }
-}
+
+  def handleRequest(sources: IncomeSourceDetailsModel, isAgent: Boolean, backUrl: String)
+                   (implicit user: MtdItUser[_]): Future[Result] = {
+    Future.successful(
```

Comment:
if you can wrap only redirect with Future.successful
on line 79: please use Future { .. }